tests/behat/features/bootstrap/EmailContext.php

```diff
@@ -223,30 +187,248 @@ public function iShouldHaveAnEmailWithTitleAndBodyMulti($subject, TableNode $tab
 
   /**
    * I do not have an email with multiple content.
    *
    * @Then I should not have an email with subject :arg1 and in the content:
    */
-  public function iShouldNotHaveAnEmailWithTitleAndBodyMulti($subject, TableNode $table) {
+  public function iShouldNotHaveAnEmailWithTitleAndBodyMulti(string $subject, TableNode $table) : void {
     $body = [];
     $hash = $table->getHash();
     foreach ($hash as $row) {
       $body[] = $row['content'];
     }
 
-    $found_email = $this->findSubjectAndBody($subject, $body);
+    $this->assertNoEmailWithSubjectAndBody($subject, $body);
+  }
+
+  /**
+   * Ensure at least one email exists matching the provided subject and body.
+   *
+   * @param string $subject
+   *   The exact subject the email should have.
+   *
+   * @param list<string> $expected_lines
+   *   A list of lines that should all be present in the body of the e-mail.
+   *
+   * @throws \Exception
+   *   An exception that details to the developer what was wrong (e.g. no
+   *   matching subject, no matching lines, or missing a specific line match).
+   */
+  protected function assertEmailWithSubjectAndBody(string $subject, array $expected_lines) : void {
+    $emails = $this->findEmailsWithSubject($subject);
+
+    // If no emails with the subject exist then we want to report that so a
+    // developer knows that may be the cause of their failure.
+    $subject_match_count = count($emails);
+    if ($subject_match_count === 0) {
+      throw new \Exception("No emails with subject '$subject' were found.");
+    }
+
+    $partial_matches = [];
+    $count_expected = count($expected_lines);
+    foreach ($emails as $email) {
+      $matched_lines = $this->getMatchingLinesForEmail($expected_lines, $email);
+
+      $count_matched = count($matched_lines);
+      // If we have a complete match then we're done.
+      if ($count_matched === $count_expected) {
+        return;
+      }
+
+      if ($count_matched > 0) {
+        $partial_matches[] = $matched_lines;
+      }
+    }
+
+    $partial_match_count = count($partial_matches);
+    // If we had no partial matches then we can provide a simplified error message.
+    if ($partial_match_count === 0) {
+      $message = $subject_match_count === 1
+        ? "One email with the subject '$subject' was found but none of the expected lines were found in the body of the email."
+        : "$subject_match_count emails with the subject '$subject' were found but none of the expected lines were found in the body of the email.";
+      throw new \Exception($message);
+    }
+    // If we had a single partial match then we tell the developer and there's
+    // probably a simple typo in the email or test.
+    if ($partial_match_count === 1) {
+      $missing_lines = array_diff($expected_lines, $partial_matches[0]);
+      $message = "One email was found which matched the subject and some lines but the following lines were missing from the e-mail: \n  " . implode("\n  ", $missing_lines);
```

Comment:
Isn't be better use PHP_EOL than "\n"

Reply:
I think you are right, but think we need to put this comment then on https://github.com/goalgorilla/open_social/pull/3500 as mine only uses this PR, but does not introduce the code itself. (although it looks like it as the PR 3500 isn't merged yet)

Reply:
Hey @vcsvinicius :D  As author of 3500 I can reply to this :) 

I think in general with respect to PHP best practices you're right. In practice though I think the code we write for Behat is only ever executed on Mac or Linux which both use `\n` as line separater, only Windows deviates with `\r\n` which is where the `PHP_EOL` would help (though I'm not sure whether `\n` alone is perhaps enough for Windows these days).

I think there's two reasons why I wouldn't use PHP_EOL here, let me know whether you (dis)agree :D 

1. Adding PHP_EOL into a string makes the code more complicated because string concatenation needs to be used and I can't use `{$variable_interpolation}`.
2. In case the error message needs to be processed by an automated process, consistent linebreaks might be slightly easier to parse than platform-specific output (unless the consuming app also properly handles different platform's linebreaks like PHP does).

Reply:
I agree, this appointment is more best practice than a trouble.